### PR TITLE
[flang][AMDGPU] Convert math ops to AMD GPU library calls instead of libm calls

### PR DESCRIPTION
**Pull request by jsjodin**
This patch invokes a pass when compiling for an AMDGPU target to lower math operations to AMD GPU library calls library calls instead of libm calls.

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
I don't think this is the way the system should work, but I guess it's reusing an existing pass

**Comment by jsjodin** [comment]
> I don't think this is the way the system should work, but I guess it's reusing an existing pass

I haven't found a better way to do this with what exists right now. The problem is that the math ops get converted to LLVM intrinsic calls in MLIR which may not be supported by the backend and there's no mechanism to map intrinsics to functions that implement them right now. There is a discussion thread on discourse, and the main proposal requires changes to the library where an attribute would be added to a function to indicate it implements a certain intrinsic.

**Comment by arsenm** [comment]
> > I don't think this is the way the system should work, but I guess it's reusing an existing pass
> 
> I haven't found a better way to do this with what exists right now. The problem is that the math ops get converted to LLVM intrinsic calls in MLIR which may not be supported by the backend and there's no mechanism to map intrinsics to functions that implement them right now. There is a discussion thread on discourse, and the main proposal requires changes to the library where an attribute would be added to a function to indicate it implements a certain intrinsic.

Right, we don't have a real platform definition for how this is supposed to work. Every frontend stitches all of these pieces together slightly differently, and the result is a big mess

File: flang/test/Lower/OpenMP/math-amdgpu.f90 [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+!$omp declare target
+  real(8) :: x, y
+!CHECK: call double @__ocml_sin_f64(double {{.*}})
+  y = sin(x)
+end subroutine omp_sin_f64
+
+subroutine omp_abs_f32(x, y)
+!$omp declare target
+  real :: x, y
+!CHECK: call float @__ocml_fabs_f32(float {{.*}})
```
Comment (arsenm) [inline flang/test/Lower/OpenMP/math-amdgpu.f90:35]: This one should definitely just go to llvm.fabs

File: flang/test/Lower/OpenMP/math-amdgpu.f90 [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  y = abs(x)
+end subroutine omp_abs_f32
+
+subroutine omp_abs_f64(x, y)
+!$omp declare target
+  real(8) :: x, y
+!CHECK: call double @__ocml_fabs_f64(double {{.*}})
```
Comment (arsenm) [inline flang/test/Lower/OpenMP/math-amdgpu.f90:42]: Same. OCML shouldn't even provide these

File: flang/test/Lower/OpenMP/math-amdgpu.f90 [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+!$omp declare target
+  real(8) :: x, y
+!CHECK: call double @__ocml_erf_f64(double {{.*}})
+  y = erf(x)
+end subroutine omp_erf_f64
+
+subroutine omp_exp_f32(x, y)
+!$omp declare target
+  real :: x, y
+!CHECK: call float @__ocml_exp_f32(float {{.*}})
```
Comment (arsenm) [inline flang/test/Lower/OpenMP/math-amdgpu.f90:105]: This should just call the llvm intrinsic

File: flang/test/Lower/OpenMP/math-amdgpu.f90 [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+!$omp declare target
+  real(8) :: x, y
+!CHECK: call double @__ocml_exp_f64(double {{.*}})
+  y = exp(x)
+end subroutine omp_exp_f64
+
+subroutine omp_log_f32(x, y)
+!$omp declare target
+  real :: x, y
+!CHECK: call float @__ocml_log_f32(float {{.*}})
```
Comment (arsenm) [inline flang/test/Lower/OpenMP/math-amdgpu.f90:119]: Should just use the llvm intrinsic

File: flang/test/Lower/OpenMP/math-amdgpu.f90 [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+!$omp declare target
+  real(8) :: x, y
+!CHECK: call double @__ocml_log_f64(double {{.*}})
+  y = log(x)
+end subroutine omp_log_f64
+
+subroutine omp_log10_f32(x, y)
+!$omp declare target
+  real :: x, y
+!CHECK: call float @__ocml_log10_f32(float {{.*}})
```
Comment (arsenm) [inline flang/test/Lower/OpenMP/math-amdgpu.f90:133]: Should just use the llvm intrinsic

File: flang/test/Lower/OpenMP/math-amdgpu.f90 [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+!$omp declare target
+  real :: x, y
+!CHECK: call float @__ocml_sqrt_f32(float {{.*}})
+  y = sqrt(x)
+end subroutine omp_sqrt_f32
+
+subroutine omp_sqrt_f64(x, y)
+!$omp declare target
+  real(8) :: x, y
+!CHECK: call double @__ocml_sqrt_f64(double {{.*}})
```
Comment (arsenm) [inline flang/test/Lower/OpenMP/math-amdgpu.f90:154]: Should just use the llvm intrinsic

File: flang/test/Lower/OpenMP/math-amdgpu.f90 [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+!$omp declare target
+  real(8) :: x, y
+!CHECK: call double @__ocml_log10_f64(double {{.*}})
+  y = log10(x)
+end subroutine omp_log10_f64
+
+subroutine omp_sqrt_f32(x, y)
+!$omp declare target
+  real :: x, y
+!CHECK: call float @__ocml_sqrt_f32(float {{.*}})
```
Comment (arsenm) [inline flang/test/Lower/OpenMP/math-amdgpu.f90:147]: Should just use the llvm intrinsic

File: flang/test/Lower/OpenMP/math-amdgpu.f90 (hunk 1 above)
Comment (krzysz00) [reply flang/test/Lower/OpenMP/math-amdgpu.f90:35]: ... then why is there an OCML function for it? Backwards compatibility?

File: flang/test/Lower/OpenMP/math-amdgpu.f90 (hunk 3 above)
Comment (krzysz00) [reply flang/test/Lower/OpenMP/math-amdgpu.f90:105]: ... Yeah, that one does seem to just wrap the intrinsic going off of an `llvm-dis`

**Review by krzysz00 — COMMENTED** [review]
@arsenm  Overall, is there a problem with using the OCML functions uniformly, even if they're trivial? That leaves the ability for them to become non-trivial in the future

**Comment by arsenm** [comment]
> @arsenm Overall, is there a problem with using the OCML functions uniformly, even if they're trivial?

Yes, especially when they're trivial. The cases that just map to an llvm intrinsic add system complexity for no benefit.

> That leaves the ability for them to become non-trivial in the future

Fundamentally we should be treating it more as a runtime library for the compiler and not a user facing abstraction. Right now it's trying to act like a user library or libm and compiler-rt all at the same time when they have different/incompatible constraints.

**Comment by jsjodin** [comment]
> > @arsenm Overall, is there a problem with using the OCML functions uniformly, even if they're trivial?
> 
> Yes, especially when they're trivial. The cases that just map to an llvm intrinsic add system complexity for no benefit.
> 

Having a more uniform interface reduces complexity. That interface should be the llvm intrinsics (or maybe libm calls), but there are holes and forcing the higher level transforms to know about what intrinsics that are implemented and not in the backend increases complexity in my opinion.

> > That leaves the ability for them to become non-trivial in the future
> 
> Fundamentally we should be treating it more as a runtime library for the compiler and not a user facing abstraction. Right now it's trying to act like a user library or libm and compiler-rt all at the same time when they have different/incompatible constraints. 

If I understand this correctly the main issue is that clang and flang use OCML at all, and it should be purely internal to LLVM.
OCML is only used within the LLVM projects from what I can tell. It doesn't seem to be user-facing, since it is not mentioned in the ROCm documentation anymore (not for a long time), so we can perhaps assume we control where and how it is used. Once a real solution is implemented we should be able to just delete this pass and let the standard lowering passes do the work. In the mean time I think it is reasonable to use OCML as the uniform interface.

What are your opinions on adding the "implements" attribute as long term solution? This would add an attribute to tell the compiler that a function implements an intrinsic something like __attribute__ implements(llvm.pow). I was looking into this, but I don't have time to work on it right now. The discussion thread is here:
https://discourse.llvm.org/t/nvptx-codegen-for-llvm-sin-and-friends/58170/33

**Comment by arsenm** [comment]
> If I understand this correctly the main issue is that clang and flang use OCML at all, and it should be purely internal to LLVM. 

Mainly yes.

> OCML is only used within the LLVM projects from what I can tell. It doesn't seem to be user-facing, since it is not mentioned in the ROCm documentation anymore (not for a long time), so we can perhaps assume we control where and how it is used. Once a real solution is implemented we should be able to just delete this pass and let the standard lowering passes do the work. In the mean time I think it is reasonable to use OCML as the uniform interface.

LLVM is the interface. OCML is an implementation detail.

> 
> What are your opinions on adding the "implements" attribute as long term solution? This would add an attribute to tell the compiler that a function implements an intrinsic something like **attribute** implements(llvm.pow). I was looking into this, but I don't have time to work on it right now. The discussion thread is here: https://discourse.llvm.org/t/nvptx-codegen-for-llvm-sin-and-friends/58170/33

It's not really how I wanted this to go, but the current trend is to have all of the math library functions available as intrinsics. Given that, I don't see much point in adding such an attribute. I was leaning towards removing any intrinsics that no backend can reasonably implement without a runtime call, and that would be complemented by improved libcall-by-name handling. It would make more sense to add implements if we went in the other direction. With this trend, I think it's jumping through hoops to maintain implementation detail names we control and can replace.

We lack a proper platform definition. I think it would be best if we defined amdhsa like a normal operating system, with a well defined set of provided libm functions, using the standard names. The __ocml prefixes are a relic from how OpenCL was implemented long ago. It would make more sense to put any extension functions in an __amd or __amdhsa prefix. Another issue is amdpal and the other triples we use. Different projects have thrown assorted builtin libraries together in different ways, so it would be better to define this bottom up.

**Comment by jsjodin** [comment]
@arsenm, if I change the conversion pass to map to the supported llvm intinsics, which you pointed out, would that be acceptable for now?

**Comment by arsenm** [comment]
> @arsenm, if I change the conversion pass to map to the supported llvm intinsics, which you pointed out, would that be acceptable for now?

Yes. Hopefully in the future we will have a sensible intrinsic lowering solution

**Review by krzysz00 — APPROVED** [review]
Seems fine to me and I'm trusting you're happy with the test coverage

**Pull request merged**